server/src/main/java/org/elasticsearch/client/node/NodeClient.java

```diff
@@ -61,25 +77,59 @@ public NodeClient(Settings settings, ThreadPool threadPool) {
     }
 
     @SuppressWarnings("rawtypes")
-    public void initialize(Map<ActionType, TransportAction> actions, TaskManager taskManager, Supplier<String> localNodeId,
+    public void initialize(Map<ActionType, TransportAction> actions, TaskManager taskManager, Supplier<DiscoveryNode> localNode,
                            RemoteClusterService remoteClusterService, NamedWriteableRegistry namedWriteableRegistry) {
         this.actions = actions;
         this.taskManager = taskManager;
-        this.localNodeId = localNodeId;
+        this.localNode = localNode;
         this.remoteClusterService = remoteClusterService;
         this.namedWriteableRegistry = namedWriteableRegistry;
     }
 
+    @SuppressWarnings({"rawtypes", "unchecked"})
     @Override
     public void close() {
-        // nothing really to do
+        if (closed.compareAndSet(false, true)) {
+            final List listenersToFail = new ArrayList<>();
+            for (Iterator<ActionListener<? extends ActionResponse>> iterator = outstandingListeners.iterator(); iterator.hasNext(); ) {
+                listenersToFail.add(iterator.next());
+                iterator.remove();
+            }
+            if (listenersToFail.isEmpty() == false) {
```

Comment:
This is a little awkward and mainly motivated by the fact that we have at least one test suit that closes the client before initializing it. I didn't want to make closing before initializing fail every time (which it would because `localNode` wouldn't be available so I added the check for empty here.
We copy while removing from the iterator instead of using `outstandingListeners.isEmpty()` because that method isn't reliable for CHM according to the docs. Also, we want to clear out the map to not leak memory for listeners which we are guaranteed to do together with https://github.com/elastic/elasticsearch/pull/62919/files#diff-532ff175220b3d116860882dc8cfc137R125 since the iterator will represent a point in time view of the map/set from after closed was set to `true`.